Mathlib/AlgebraicGeometry/Stalk.lean

```diff
@@ -51,12 +61,264 @@ theorem IsAffineOpen.fromSpecStalk_eq {X : Scheme} (x : X) {U V : X.Opens}
+noncomputable
+def stalkClosedPointIso :
+    (Spec R).presheaf.stalk (closedPoint R) ≅ R :=
+  StructureSheaf.stalkIso _ _ ≪≫ (IsLocalization.atUnits R
+      (closedPoint R).asIdeal.primeCompl fun _ ↦ not_not.mp).toRingEquiv.toCommRingCatIso.symm
+
+lemma stalkClosedPointIso_inv :
+    (stalkClosedPointIso R).inv = StructureSheaf.toStalk R _ := by
+  ext x
+  exact StructureSheaf.localizationToStalk_of _ _ _
+
+lemma ΓSpecIso_hom_stalkClosedPointIso_inv :
+    (Scheme.ΓSpecIso R).hom ≫ (stalkClosedPointIso R).inv =
+      (Spec R).presheaf.germ ⊤ (closedPoint _) trivial := by
+  rw [stalkClosedPointIso_inv, ← Iso.eq_inv_comp]
+  rfl
+
+@[reassoc (attr := simp)]
+lemma germ_stalkClosedPointIso_hom :
+    (Spec R).presheaf.germ ⊤ (closedPoint _) trivial ≫ (stalkClosedPointIso R).hom =
+      (Scheme.ΓSpecIso R).hom := by
+  rw [← ΓSpecIso_hom_stalkClosedPointIso_inv, Category.assoc, Iso.inv_hom_id, Category.comp_id]
+
+lemma Spec_stalkClosedPointIso :
+    Spec.map (stalkClosedPointIso R).inv = (Spec R).fromSpecStalk (closedPoint R) := by
+  rw [stalkClosedPointIso_inv, Scheme.Spec_fromSpecStalk']
+
+end stalkClosedPointIso
+
+section stalkClosedPointTo
+
+variable {R} (f : Spec R ⟶ X)
+
+-- move me
+lemma LocalRing.closed_point_mem_iff {R : Type*} [CommRing R] [LocalRing R]
+    {U : Opens (PrimeSpectrum R)} : closedPoint R ∈ U ↔ U = ⊤ :=
+  ⟨(eq_top_iff.mpr fun x _ ↦ (specializes_closedPoint x).mem_open U.2 ·), (· ▸ trivial)⟩
+
+@[simp]
+lemma Spec_closedPoint {R S : CommRingCat} [LocalRing R] [LocalRing S]
+    {f : R ⟶ S} [IsLocalRingHom f] : (Spec.map f).1.base (closedPoint S) = closedPoint R :=
+  LocalRing.comap_closedPoint f
+
+namespace Scheme
+
+/--
+Given a local ring `(R, 𝔪)` and a morphism `f : Spec R ⟶ X`,
+they induce a ring homomorphim `φ : 𝒪_{X, f 𝔪} ⟶ X`.
+
+This is inverse to `φ ↦ Spec.map φ ≫ X.fromSpecStalk (f 𝔪)`. See `SpecToEquivOfLocalRing`.
+-/
+noncomputable
+def stalkClosedPointTo :
+    X.presheaf.stalk (f.1.base (closedPoint R)) ⟶ R :=
+  f.stalkMap (closedPoint R) ≫ (stalkClosedPointIso R).hom
+
+instance isLocalRingHom_stalkClosedPointTo :
+    IsLocalRingHom (stalkClosedPointTo f) := by
+  apply (config := { allowSynthFailures := true }) RingHom.isLocalRingHom_comp
+  · apply isLocalRingHom_of_iso
+  · apply f.2
+
+lemma preimage_eq_top_of_closedPoint_mem
+    {U : Opens X} (hU : f.1.base (closedPoint R) ∈ U) : f ⁻¹ᵁ U = ⊤ :=
+  LocalRing.closed_point_mem_iff.mp hU
+
+lemma stalkClosedPointTo_comp (g : X ⟶ Y) :
+    stalkClosedPointTo (f ≫ g) = g.stalkMap _ ≫ stalkClosedPointTo f := by
+  rw [stalkClosedPointTo, Scheme.stalkMap_comp]
+  exact Category.assoc _ _ _
+
+lemma germ_stalkClosedPointTo_Spec {R S : CommRingCat} [LocalRing S] (φ : R ⟶ S):
+    (Spec R).presheaf.germ ⊤ _ trivial ≫ stalkClosedPointTo (Spec.map φ) =
+      (ΓSpecIso R).hom ≫ φ := by
+  rw [stalkClosedPointTo, Scheme.stalkMap_germ_assoc, ← Iso.inv_comp_eq,
+    ← ΓSpecIso_inv_naturality_assoc]
+  erw [germ_stalkClosedPointIso_hom]
```

Comment:
We can get rid of this `erw` by adding a lemma
```
lemma preimage_top (f : X ⟶ Y) : f ⁻¹ᵁ ⊤ = ⊤ :=
  rfl
```

Reply:
It seems to throw a motive incorrect error for me.

Reply:
This
```
lemma germ_stalkClosedPointTo_Spec {R S : CommRingCat} [LocalRing S] (φ : R ⟶ S):
    (Spec R).presheaf.germ ⊤ _ trivial ≫ stalkClosedPointTo (Spec.map φ) =
      (ΓSpecIso R).hom ≫ φ := by
  rw [stalkClosedPointTo, Scheme.stalkMap_germ_assoc, ← Iso.inv_comp_eq,
    ← ΓSpecIso_inv_naturality_assoc]
  simp_rw [preimage_top]
  rw [germ_stalkClosedPointIso_hom]
  rw [Iso.inv_hom_id, Category.comp_id]
```
works for me.

Reply:
Oh `simp_rw`. Makes sense.